packages/flutter_tools/lib/src/base/logger.dart

```diff
@@ -395,123 +422,154 @@ enum _LogType { error, status, trace }
+///
 /// Generally, consider `logger.startProgress` instead of directly creating
 /// a [Status] or one of its subclasses.
-class Status {
-  Status({this.onFinish});
+abstract class Status {
+  Status({ @required this.timeout, this.onFinish }) : assert(timeout != null);
 
-  /// A straight [Status] or an [AnsiSpinner] (depending on whether the
+  /// A [SilentStatus] or an [AnsiSpinner] (depending on whether the
   /// terminal is fancy enough), already started.
-  factory Status.withSpinner({ VoidCallback onFinish }) {
+  factory Status.withSpinner({ @required Duration timeout, VoidCallback onFinish }) {
+    assert(timeout != null);
     if (terminal.supportsColor)
-      return AnsiSpinner(onFinish: onFinish)..start();
-    return Status(onFinish: onFinish)..start();
+      return AnsiSpinner(timeout: timeout, onFinish: onFinish)..start();
+    return SilentStatus(timeout: timeout, onFinish: onFinish)..start();
   }
 
+  final Duration timeout;
   final VoidCallback onFinish;
 
   bool _isStarted = false;
 
+  @protected
+  final Stopwatch stopwatch = Stopwatch();
+
+  @protected
+  bool get seemsSlow => stopwatch.elapsed > timeout;
+
+  @protected
+  String get elapsedTime {
+    if (timeout > fastOperation)
+      return getElapsedAsSeconds(stopwatch.elapsed);
+    return getElapsedAsMilliseconds(stopwatch.elapsed);
+  }
+
   /// Call to start spinning.
   void start() {
     assert(!_isStarted);
+    assert(!stopwatch.isRunning);
+    stopwatch.start();
     _isStarted = true;
   }
 
   /// Call to stop spinning after success.
   void stop() {
-    assert(_isStarted);
-    _isStarted = false;
-    if (onFinish != null)
-      onFinish();
+    finish();
   }
 
   /// Call to cancel the spinner after failure or cancellation.
   void cancel() {
+    finish();
+  }
+
+  @protected
+  void finish() {
     assert(_isStarted);
+    assert(stopwatch.isRunning);
+    stopwatch.stop();
     _isStarted = false;
     if (onFinish != null)
       onFinish();
   }
 }
 
+/// A [SilentStatus] shows nothing.
+class SilentStatus extends Status {
+  SilentStatus({
+    @required Duration timeout,
+    VoidCallback onFinish,
+  }) : super(timeout: timeout, onFinish: onFinish);
+}
+
 /// An [AnsiSpinner] is a simple animation that does nothing but implement a
-/// ASCII/Unicode spinner. When stopped or canceled, the animation erases
-/// itself.
+/// terminal spinner. When stopped or canceled, the animation erases itself.
 class AnsiSpinner extends Status {
-  AnsiSpinner({VoidCallback onFinish}) : super(onFinish: onFinish);
+  AnsiSpinner({
+    @required Duration timeout,
+    VoidCallback onFinish,
+  }) : super(timeout: timeout, onFinish: onFinish);
 
   int ticks = 0;
   Timer timer;
 
+  bool _showingSlowWarning = false;
+
   // Windows console font has a limited set of Unicode characters.
   List<String> get _animation => platform.isWindows
       ? <String>[r'-', r'\', r'|', r'/']
       : <String>['⣾', '⣽', '⣻', '⢿', '⡿', '⣟', '⣯', '⣷'];
```

Comment:
Is anyone else bothered by the fact that this runs counter-clockwise? 😄

Reply:
yes, but greg likes it. :-)